rtl838x: Add switch and ethernet driver support for RTL839x SoCs

bkobl opened the pull request:
This adds support for RTL839x SoCs in the ethernet and switch
drivers of the rtl838x architecture.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

mtdcr commented on target/linux/rtl838x/files-5.4/drivers/net/dsa/rtl838x_sw.c:
```diff
@@ -1637,13 +2090,10 @@ static int rtl838x_port_enable(struct dsa_switch *ds, int port,
 
 	/* add port to switch mask of CPU_PORT */
-	priv->r->mask_port_iso_ctrl(0, 1 << port, priv->cpu_port);
+	priv->r->mask_port_reg_be(0ULL, 1ULL << port, priv->r->port_iso_ctrl(priv->cpu_port));
```
You can keep the `0`.

bkobl replied:
done.

mtdcr commented on target/linux/rtl838x/files-5.4/drivers/net/ethernet/rtl838x_eth.c:
```diff
@@ -448,34 +683,27 @@ static int rtl838x_eth_tx(struct sk_buff *skb, struct net_device *dev)
 		memcpy((void *)KSEG1ADDR(h->buf), skb->data, len);
 		/* Make sure packet data is visible to ASIC */
-		mb();
+		mb(); /* wmb() probably works, too */
```
Maybe use `memcpy_toio()` here.

bkobl replied:
memcpy_toio would not do the subsequent memory barrier.

mtdcr reviewed (changes requested):
The patch is quite big, which made reviewing difficult. See my review comments inline.

bkobl commented on target/linux/rtl838x/files-5.4/drivers/net/dsa/rtl838x.h:
```diff
@@ -160,10 +221,26 @@ struct rtl838x_reg {
 	int imr_glb;
 	void (*vlan_tables_read)(u32 vlan, struct rtl838x_vlan_info *info);
-	void (*vlan_set_tagged)(u32 vlan, u64 portmask, u32 conf);
+	void (*vlan_set_tagged)(u32 vlan, struct rtl838x_vlan_info *info);
```
Will fix.

bkobl commented on target/linux/rtl838x/files-5.4/drivers/net/ethernet/rtl838x_eth.c:
```diff
@@ -174,36 +285,79 @@ static irqreturn_t rtl838x_net_irq(int irq, void *dev_id)
 		rtl838x_rb_cleanup(priv);
 	}
 
+	if (priv->family_id == RTL8390_FAMILY_ID && status & 0x00100000) {
+		sw_w32(0x00100000, priv->r->dma_if_intr_sts);
+		rtl839x_l2_notification_handler(priv);
+	}
+
+	if (priv->family_id == RTL8390_FAMILY_ID && status & 0x00200000) {
+		sw_w32(0x00200000, priv->r->dma_if_intr_sts);
+		rtl839x_l2_notification_handler(priv);
+	}
+
+	if (priv->family_id == RTL8390_FAMILY_ID && status & 0x00400000) {
+		sw_w32(0x00400000, priv->r->dma_if_intr_sts);
+		rtl839x_l2_notification_handler(priv);
+	}
```
Thanks, will fix!

bkobl commented on target/linux/rtl838x/files-5.4/drivers/net/dsa/rtl838x_sw.c:
```diff
@@ -21,8 +22,12 @@
+
 DEFINE_MUTEX(smi_lock);
 
+extern void rtl8380_sds_rst(int mac);
```
Will fix.

bkobl commented on target/linux/rtl838x/files-5.4/drivers/net/dsa/rtl838x_sw.c:
```diff
@@ -182,29 +247,42 @@ static void rtl838x_vlan_tables_read(u32 vlan, struct rtl838x_vlan_info *info)
 }
 
-static void rtl839x_vlan_set_tagged(u32 vlan, u64 portmask, u32 conf)
+static void rtl839x_vlan_set_tagged(u32 vlan, struct rtl838x_vlan_info *info)
```
Thanks for the suggestion, will do!

bkobl replied:
Will fix.

bkobl commented on target/linux/rtl838x/files-5.4/drivers/net/dsa/rtl838x_sw.c:
```diff
@@ -133,10 +146,51 @@ inline int rtl839x_tbl_access_data_0(int i)
+inline int rtl838x_vlan_profile(int profile)
+{
+	return RTL838X_VLAN_PROFILE(profile);
+}
+
+inline int rtl839x_vlan_profile(int profile)
+{
+	return RTL839X_VLAN_PROFILE(profile);
+}
+
+inline int rtl838x_vlan_port_egr_filter(int port)
+{
+	return RTL838X_VLAN_PORT_EGR_FLTR;
+}
+
+inline int rtl839x_vlan_port_egr_filter(int port)
+{
+	return RTL839X_VLAN_PORT_EGR_FLTR(port);
+}
+
+inline int rtl838x_vlan_port_igr_filter(int port)
+{
+	return RTL838X_VLAN_PORT_IGR_FLTR(port);
+}
+
+inline int rtl839x_vlan_port_igr_filter(int port)
+{
+	return RTL839X_VLAN_PORT_IGR_FLTR(port);
+}
+
+inline int rtl838x_vlan_port_pb(int port)
+{
+	return RTL838X_VLAN_PORT_PB_VLAN(port);
+}
+
+inline int rtl839x_vlan_port_pb(int port)
+{
+	return RTL839X_VLAN_PORT_PB_VLAN(port);
+}
+
```
I will look into this, to clean it up. It In the given cases it is probably cleaner to use inline functions in which the register offset is calculated.

bkobl commented on target/linux/rtl838x/files-5.4/drivers/net/dsa/rtl838x_sw.c:
```diff
@@ -182,29 +247,42 @@ static void rtl838x_vlan_tables_read(u32 vlan, struct rtl838x_vlan_info *info)
 }
 
-static void rtl838x_vlan_set_tagged(u32 vlan, u64 portmask, u32 conf)
+static void rtl838x_vlan_set_tagged(u32 vlan, struct rtl838x_vlan_info *info)
```
Will fix.

bkobl commented on target/linux/rtl838x/files-5.4/drivers/net/dsa/rtl838x_sw.c:
```diff
@@ -238,15 +316,272 @@ static inline int rtl839x_mac_force_mode_ctrl(int p)
+		| (idx & 0x1fff);
+
+	sw_w32(cmd, RTL838X_TBL_ACCESS_L2_CTRL);
+	do { }  while (sw_r32(RTL838X_TBL_ACCESS_L2_CTRL) & (1 << 16));
```
The delay is extremely short, not more than a few CPU cycles, since it is merely waiting for the switch-engine which is implemented as an ASIC to load some registers.

adschm commented:
@mtdcr If you are fine with the changes, consider providing a Reviewed-by: eventually; this might increase the chance of the PR being merged, or at least might have it merged quicker.

bkobl commented on target/linux/rtl838x/files-5.4/drivers/net/dsa/rtl838x_sw.c:
```diff
@@ -238,15 +316,272 @@ static inline int rtl839x_mac_force_mode_ctrl(int p)
+		e->type = IP6_MULTICAST;
+	}
+
+	entry = (((u64) r[0]) << 12) | ((r[1] & 0xfffffff0) << 12) | ((r[2] >> 4) & 0xfff);
```
Indeed, the first 12 should be a 44. I also fixed the overflow.

bkobl commented on target/linux/rtl838x/files-5.4/drivers/net/dsa/rtl838x.h:
```diff
@@ -133,20 +163,51 @@ struct rtl838x_port {
+};
+
+struct rtl838x_l2_entry {
+	u8 mac[6];
```
Fixed.

Commit:
rtl838x: Add switch and ethernet driver support for RTL839x SoCs

This adds support for RTL839x SoCs in the ethernet and switch
drivers of the rtl838x architecture.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

bkobl commented:
I believe I have addressed all the remarks in the commit I just pushed.

mtdcr commented on target/linux/rtl838x/files-5.4/drivers/net/dsa/rtl838x_sw.c:
```diff
@@ -30,34 +31,46 @@ struct rtl838x_mib_desc {
 };
 
-inline void rtl838x_mask_port_reg(u64 clear, u64 set, int reg)
+static inline void rtl838x_mask_port_reg(u64 clear, u64 set, int reg)
```
This function (and many other functions below) gets referenced by a function pointer and therefore can't get inlined. Please remove the `inline` keyword.

mtdcr commented on target/linux/rtl838x/files-5.4/drivers/net/dsa/rtl838x_sw.c:
```diff
@@ -30,34 +31,46 @@ struct rtl838x_mib_desc {
-inline void rtl838x_mask_port_reg(u64 clear, u64 set, int reg)
+static inline void rtl838x_mask_port_reg(u64 clear, u64 set, int reg)
 {
 	sw_w32_mask((u32)clear, (u32)set, reg);
```
Because `sw_*` are implemented as macros, it's not obvious, but these u32-casts don't have any effect.

I suggest reimplementing `sw_*` as inline functions in a follow-up PR to make clear which types are expected.

mtdcr commented on target/linux/rtl838x/files-5.4/drivers/net/dsa/rtl838x_sw.c:
```diff
@@ -30,34 +31,46 @@ struct rtl838x_mib_desc {
+	return ((u64) sw_r32(reg));
+}
+
+static inline void rtl839x_mask_port_reg_be(u64 clear, u64 set, int reg)
+{
+	sw_w32_mask((u32)(clear >> 32), (u32)(set >> 32), reg);
+	sw_w32_mask((u32)(clear & 0xffffffff), (u32)(set & 0xffffffff), reg + 4);
```
Maybe you could use something built upon __raw_readq() and __raw_writeq() here.

mtdcr reviewed (approved):
Looks good to me. Feel free to address my new comments in a later patch.

ynezz commented:
@mtdcr Thanks for the review, feel free to provide `Reviewed-by:` tag.

mtdcr commented:
`Reviewed-by: Andreas Oberritter <obi@saftware.de>`

ynezz commented:
Thanks! Pulled into my staging tree at https://git.openwrt.org/openwrt/staging/ynezz.git

Neustradamus commented:
Merged commit:
- https://github.com/openwrt/openwrt/commit/fba69a5a3086d37adb3816ab65f2bf372cd099e0
